src/Processors/QueryPlan/ReadFromMergeTree.cpp

```diff
@@ -824,11 +829,14 @@ Pipe ReadFromMergeTree::spreadMarkRangesAmongStreamsFinal(
             pipe.addSimpleTransform([sorting_expr](const Block & header)
                                     { return std::make_shared<ExpressionTransform>(header, sorting_expr); });
 
-        /// If do_not_merge_across_partitions_select_final is true and there is only one part in partition
-        /// with level > 0 then we won't postprocess this part
+        /// If do_not_merge_across_partitions_select_final is true
+        /// and there is only one part in partition with level > 0
+        /// then we won't postprocess this part
+        bool should_read_in_order = !input_order_info || input_order_info->direction == 0;
```

Comment:
Why should we read in order if input_order_info is null

Reply:
My mistake was that it should be `should_not_read_in_order`.
But anyway, decided to use another solution. We need to disable read-in-order optimization for reverse order with `final` because it may lead to incorrect `final` behavior (it may rely on read in direct order). Check out new commits in the PR, please.